File: src/storage/cfs.ts
import { createHash } from 'node:crypto';
import { constants, type BigIntStats } from 'node:fs';
import { chmod, link, lstat, lutimes, mkdir, open, readFile, readdir, readlink, rename, rm, stat, symlink, utimes } from 'node:fs/promises';
import { dirname, join, relative, resolve, sep } from 'node:path';
import type { Digest } from '../domain/types.ts';
import { ServiceError } from '../domain/errors.ts';

const MAGIC = Buffer.from('CFS-v1\0', 'utf8');
const MAX_FILE_BYTES = 256 * 1024 * 1024;
const MAX_TREE_BYTES = 1024 * 1024 * 1024;
const TYPE_DIRECTORY = 1;
const TYPE_FILE = 2;
const TYPE_SYMLINK = 3;
const TYPE_HARDLINK = 4;

interface EntryHeader {
  path: string;
  type: number;
  mode: number;
  mtimeNs: bigint;
  payloadLength: number;
}

export interface CfsObject {
  treeDigest: Digest;
  path: string;
  sizeBytes: number;
}

function u32(value: number): Buffer {
  const out = Buffer.allocUnsafe(4);
  out.writeUInt32BE(value);
  return out;
}

function u64(value: bigint): Buffer {
  const out = Buffer.allocUnsafe(8);
  out.writeBigUInt64BE(value);
  return out;
}

function digest(bytes: Uint8Array | string): Digest {
  return `sha256:${createHash('sha256').update(bytes).digest('hex')}` as Digest;
}

function validateRelativePath(path: string): void {
  if (path.length === 0 || path.includes('\0') || path.startsWith('/') || path.split('/').some((part) => part === '.' || part === '..' || part.length === 0)) {
    throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `Unsafe snapshot path: ${path}`);
  }
}

function encodeHeader(header: EntryHeader): Buffer {
  const pathBytes = Buffer.from(header.path, 'utf8');
  return Buffer.concat([
    u32(pathBytes.length),
    pathBytes,
    Buffer.from([header.type]),
    u32(header.mode),
    u64(header.mtimeNs),
    u64(BigInt(header.payloadLength)),
  ]);
}

async function collectEntries(root: string): Promise<Array<{ absolute: string; relative: string; stats: BigIntStats }>> {
  const entries: Array<{ absolute: string; relative: string; stats: BigIntStats }> = [];
  async function walk(directory: string): Promise<void> {
    const names = await readdir(directory, { encoding: 'utf8' });
    names.sort((a, b) => Buffer.compare(Buffer.from(a), Buffer.from(b)));
    for (const name of names) {
      const absolute = join(directory, name);
      const relativePath = relative(root, absolute).split(sep).join('/');
      validateRelativePath(relativePath);
      const entryStats = await lstat(absolute, { bigint: true });
      entries.push({ absolute, relative: relativePath, stats: entryStats });
      if (entryStats.isDirectory()) await walk(absolute);
    }
  }
  await walk(root);
  entries.sort((a, b) => Buffer.compare(Buffer.from(a.relative), Buffer.from(b.relative)));
  return entries;
}

export async function exportCfs(root: string, objectRoot: string): Promise<CfsObject> {
  const rootStats = await stat(root).catch(() => undefined);
  if (!rootStats?.isDirectory()) throw new ServiceError('INVALID_ARGUMENT', 'Workspace root does not exist');
  await mkdir(join(objectRoot, 'sha256'), { recursive: true, mode: 0o700 });
  const stagingDir = join(objectRoot, '..', '..', 'staging');
  await mkdir(stagingDir, { recursive: true, mode: 0o700 });
  const stagingPath = join(stagingDir, `${crypto.randomUUID()}.cfs`);
  const handle = await open(stagingPath, constants.O_CREAT | constants.O_EXCL | constants.O_WRONLY, 0o600);
  const hasher = createHash('sha256');
  let total = 0;
  const hardLinks = new Map<string, string>();
  const append = async (chunk: Uint8Array): Promise<void> => {
    total += chunk.byteLength;
    if (total > MAX_TREE_BYTES) throw new ServiceError('SNAPSHOT_LIMIT', 'Snapshot exceeds total size limit');
    hasher.update(chunk);
    await handle.write(chunk);
  };
  try {
    await append(MAGIC);
    for (const entry of await collectEntries(root)) {
      const mode = Number(entry.stats.mode) & 0o7777;
      if ((mode & 0o6000) !== 0) throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `setuid/setgid entry rejected: ${entry.relative}`);
      const mtimeNs = ((entry.stats.mtimeNs + 500_000n) / 1_000_000n) * 1_000_000n;
      if (entry.stats.isDirectory()) {
        await append(encodeHeader({ path: entry.relative, type: TYPE_DIRECTORY, mode, mtimeNs, payloadLength: 0 }));
      } else if (entry.stats.isSymbolicLink()) {
        const target = await readlink(entry.absolute, 'utf8');
        const resolvedTarget = resolve(dirname(entry.absolute), target);
        if (target.startsWith('/') || (resolvedTarget !== root && !resolvedTarget.startsWith(`${root}${sep}`))) {
          throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `Symlink target escapes workspace: ${entry.relative}`);
        }
        const payload = Buffer.from(target, 'utf8');
        await append(encodeHeader({ path: entry.relative, type: TYPE_SYMLINK, mode, mtimeNs, payloadLength: payload.length }));
        await append(payload);
      } else if (entry.stats.isFile()) {
        const size = Number(entry.stats.size);
        if (size > MAX_FILE_BYTES) throw new ServiceError('SNAPSHOT_LIMIT', `File exceeds size limit: ${entry.relative}`);
        const inodeKey = `${String(entry.stats.dev)}:${String(entry.stats.ino)}`;
        const linkTarget = entry.stats.nlink > 1n ? hardLinks.get(inodeKey) : undefined;
        if (linkTarget) {
          const payload = Buffer.from(linkTarget, 'utf8');
          await append(encodeHeader({ path: entry.relative, type: TYPE_HARDLINK, mode, mtimeNs, payloadLength: payload.length }));
          await append(payload);
        } else {
          if (entry.stats.nlink > 1n) hardLinks.set(inodeKey, entry.relative);
          await append(encodeHeader({ path: entry.relative, type: TYPE_FILE, mode, mtimeNs, payloadLength: size }));
          const input = await open(entry.absolute, constants.O_RDONLY | constants.O_NOFOLLOW);
          try {
            const buffer = Buffer.allocUnsafe(64 * 1024);
            let offset = 0;
            while (offset < size) {
              const { bytesRead } = await input.read(buffer, 0, Math.min(buffer.length, size - offset), offset);
              if (bytesRead === 0) break;
              await append(buffer.subarray(0, bytesRead));
              offset += bytesRead;
            }
            if (offset !== size) throw new ServiceError('BACKEND_ERROR', `File changed during snapshot: ${entry.relative}`);
          } finally {
            await input.close();
          }
        }
      } else {
        throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `Unsupported entry: ${entry.relative}`);
      }
    }
    await handle.sync();
    const treeDigest = `sha256:${hasher.digest('hex')}` as Digest;
    const finalPath = join(objectRoot, 'sha256', treeDigest.slice(7));
    await handle.close();
    await link(stagingPath, finalPath).catch(async (error: NodeJS.ErrnoException) => {
      if (error.code !== 'EEXIST') throw error;
    });
    await rm(stagingPath, { force: true });
    await chmod(finalPath, 0o400);
    return { treeDigest, path: finalPath, sizeBytes: total };
  } catch (error) {
    await handle.close().catch(() => undefined);
    await rm(stagingPath, { force: true });
    throw error;
  }
}

function readHeader(buffer: Buffer, offset: number): { header: EntryHeader; next: number } {
  if (offset + 4 > buffer.length) throw new ServiceError('BACKEND_ERROR', 'Truncated CFS path length');
  const pathLength = buffer.readUInt32BE(offset);
  offset += 4;
  if (offset + pathLength + 21 > buffer.length) throw new ServiceError('BACKEND_ERROR', 'Truncated CFS entry');
  const path = buffer.subarray(offset, offset + pathLength).toString('utf8');
  offset += pathLength;
  const type = buffer[offset] ?? 0;
  offset += 1;
  const mode = buffer.readUInt32BE(offset);
  offset += 4;
  const mtimeNs = buffer.readBigUInt64BE(offset);
  offset += 8;
  const payloadLengthBig = buffer.readBigUInt64BE(offset);
  offset += 8;
  if (payloadLengthBig > BigInt(Number.MAX_SAFE_INTEGER)) throw new ServiceError('SNAPSHOT_LIMIT', 'CFS payload is too large');
  if ((mode & 0o6000) !== 0) throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `setuid/setgid entry rejected: ${path}`);
  validateRelativePath(path);
  return { header: { path, type, mode, mtimeNs, payloadLength: Number(payloadLengthBig) }, next: offset };
}

function safeDestination(root: string, relativePath: string): string {
  const destination = resolve(root, relativePath);
  if (destination !== root && !destination.startsWith(`${root}${sep}`)) throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', 'CFS path escapes workspace');
  return destination;
}

function assertNoSymlinkAncestor(relativePath: string, entryTypes: Map<string, number>): void {
  const parts = relativePath.split('/');
  let ancestor = '';
  for (const part of parts.slice(0, -1)) {
    ancestor = ancestor.length === 0 ? part : `${ancestor}/${part}`;
    if (entryTypes.get(ancestor) === TYPE_SYMLINK) {
      throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `CFS path traverses symlink: ${relativePath}`);
    }
  }
}
export async function ingestCfs(sourcePath: string, objectRoot: string, expectedDigest?: Digest): Promise<CfsObject> {
  const bytes = await readFile(sourcePath);
  if (bytes.length > MAX_TREE_BYTES) throw new ServiceError('SNAPSHOT_LIMIT', 'Snapshot exceeds total size limit');
  const treeDigest = digest(bytes);
  if (expectedDigest && treeDigest !== expectedDigest) throw new ServiceError('BACKEND_ERROR', 'Guest snapshot digest mismatch');
  const validationRoot = join(objectRoot, '..', '..', 'staging', `validate-${crypto.randomUUID()}`);
  const validationWorkspace = join(validationRoot, 'workspace');
  await mkdir(validationRoot, { recursive: true, mode: 0o700 });
  try {
    await extractCfs(sourcePath, validationWorkspace);
  } finally {
    await rm(validationRoot, { recursive: true, force: true });
  }
  await mkdir(join(objectRoot, 'sha256'), { recursive: true, mode: 0o700 });
  const finalPath = join(objectRoot, 'sha256', treeDigest.slice(7));
  await link(sourcePath, finalPath).catch(async (error: NodeJS.ErrnoException) => {
    if (error.code === 'EEXIST') return;
    if (error.code !== 'EXDEV') throw error;
    const target = await open(finalPath, constants.O_CREAT | constants.O_EXCL | constants.O_WRONLY, 0o400);
    try {
      await target.write(bytes);
      await target.sync();
    } finally {
      await target.close();
    }
  });
  await chmod(finalPath, 0o400);
  return { treeDigest, path: finalPath, sizeBytes: bytes.length };
}


export async function extractCfs(objectPath: string, destination: string): Promise<void> {
  const bytes = await readFile(objectPath);
  if (!bytes.subarray(0, MAGIC.length).equals(MAGIC)) throw new ServiceError('BACKEND_ERROR', 'Invalid CFS magic');
  const parent = dirname(destination);
  const staging = join(parent, `.staging-${crypto.randomUUID()}`);
  await mkdir(staging, { recursive: false, mode: 0o700 });
  let offset = MAGIC.length;
  const entryTypes = new Map<string, number>();
  const directoryTimes: Array<{ path: string; seconds: number }> = [];
  try {
    while (offset < bytes.length) {
      const decoded = readHeader(bytes, offset);
      const header = decoded.header;
      offset = decoded.next;
      if (offset + header.payloadLength > bytes.length) throw new ServiceError('BACKEND_ERROR', 'Truncated CFS payload');
      const payload = bytes.subarray(offset, offset + header.payloadLength);
      offset += header.payloadLength;
      assertNoSymlinkAncestor(header.path, entryTypes);
      const output = safeDestination(staging, header.path);
      await mkdir(dirname(output), { recursive: true, mode: 0o700 });
      if (header.type === TYPE_DIRECTORY) {
        await mkdir(output, { recursive: false, mode: header.mode & 0o777 });
        directoryTimes.push({ path: output, seconds: Number(header.mtimeNs) / 1_000_000_000 });
      } else if (header.type === TYPE_FILE) {
        const file = await open(output, constants.O_CREAT | constants.O_EXCL | constants.O_WRONLY | constants.O_NOFOLLOW, header.mode & 0o777);
        await file.write(payload);
        await file.sync();
        await file.close();
      } else if (header.type === TYPE_SYMLINK) {
        const target = payload.toString('utf8');
        if (target.includes('\0') || target.startsWith('/')) throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `Unsafe symlink target: ${header.path}`);
        const resolvedTarget = resolve(dirname(output), target);
        if (resolvedTarget !== staging && !resolvedTarget.startsWith(`${staging}${sep}`)) throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `Symlink target escapes workspace: ${header.path}`);
        await symlink(target, output);
        const seconds = Number(header.mtimeNs) / 1_000_000_000;
        await lutimes(output, seconds, seconds);
      } else if (header.type === TYPE_HARDLINK) {
        const target = payload.toString('utf8');
        validateRelativePath(target);
        if (entryTypes.get(target) !== TYPE_FILE) throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `Hard-link target is not a prior regular file: ${header.path}`);
        await link(safeDestination(staging, target), output);
      } else {
        throw new ServiceError('SNAPSHOT_UNSUPPORTED_ENTRY', `Unsupported CFS entry type: ${header.type}`);
      }
      if (header.type !== TYPE_SYMLINK) {
        await chmod(output, header.mode & 0o777);
        if (header.type !== TYPE_DIRECTORY) {
          const seconds = Number(header.mtimeNs) / 1_000_000_000;
          await utimes(output, seconds, seconds);
        }
      }
      entryTypes.set(header.path, header.type);
    }
    for (const directory of directoryTimes.reverse()) await utimes(directory.path, directory.seconds, directory.seconds);
    await rm(destination, { recursive: true, force: true });
    await rename(staging, destination);
  } catch (error) {
    await rm(staging, { recursive: true, force: true });
    throw error;
  }
}

export async function materializeCfs(objectPath: string, destination: string): Promise<void> {
  await mkdir(dirname(destination), { recursive: true, mode: 0o700 });
  await extractCfs(objectPath, destination);
}

export function commitDigest(contractDigest: Digest, treeDigest: Digest, parentId?: Digest): Digest {
  return digest(`electrosphere-commit\0v1\0${contractDigest}\0${treeDigest}\0${parentId ?? ''}`);
}

File: agent/src/cfs.rs
use anyhow::{bail, Context, Result};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::fs::{self, File, OpenOptions};
use std::io::{Read, Write};
use std::os::unix::fs::{MetadataExt, PermissionsExt};
use std::path::{Component, Path, PathBuf};

const MAGIC: &[u8] = b"CFS-v1\0";
const MAX_FILE_BYTES: u64 = 256 * 1024 * 1024;
const MAX_TREE_BYTES: u64 = 1024 * 1024 * 1024;
const TYPE_DIRECTORY: u8 = 1;
const TYPE_FILE: u8 = 2;
const TYPE_SYMLINK: u8 = 3;
const TYPE_HARDLINK: u8 = 4;

struct SnapshotWriter {
    file: File,
    hasher: Sha256,
    size: u64,
}

impl SnapshotWriter {
    fn create(destination: &Path) -> Result<Self> {
        let file = OpenOptions::new()
            .write(true)
            .create_new(true)
            .open(destination)
            .with_context(|| format!("create {}", destination.display()))?;
        Ok(Self {
            file,
            hasher: Sha256::new(),
            size: 0,
        })
    }

    fn finish(mut self) -> Result<(u64, String)> {
        self.file.flush()?;
        self.file.sync_all()?;
        let size = self.size;
        let hash = self.hasher.finalize();
        Ok((size, format!("sha256:{hash:x}")))
    }
}

impl Write for SnapshotWriter {
    fn write(&mut self, buffer: &[u8]) -> std::io::Result<usize> {
        let next = self
            .size
            .checked_add(buffer.len() as u64)
            .ok_or_else(|| std::io::Error::other("snapshot exceeds total limit"))?;
        if next > MAX_TREE_BYTES {
            return Err(std::io::Error::other("snapshot exceeds total limit"));
        }
        self.file.write_all(buffer)?;
        self.hasher.update(buffer);
        self.size = next;
        Ok(buffer.len())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        self.file.flush()
    }
}

fn validate_relative(path: &Path) -> Result<String> {
    if path.is_absolute()
        || path
            .components()
            .any(|part| matches!(part, Component::ParentDir | Component::CurDir))
    {
        bail!("unsafe snapshot path");
    }
    let value = path.to_string_lossy().replace('\\', "/");
    if value.is_empty() || value.as_bytes().contains(&0) {
        bail!("unsafe snapshot path");
    }
    Ok(value)
}

fn write_header(
    out: &mut SnapshotWriter,
    path: &str,
    kind: u8,
    mode: u32,
    mtime_ns: u64,
    payload_len: u64,
) -> Result<()> {
    out.write_all(&(path.len() as u32).to_be_bytes())?;
    out.write_all(path.as_bytes())?;
    out.write_all(&[kind])?;
    out.write_all(&mode.to_be_bytes())?;
    out.write_all(&mtime_ns.to_be_bytes())?;
    out.write_all(&payload_len.to_be_bytes())?;
    Ok(())
}

fn collect(root: &Path, directory: &Path, entries: &mut Vec<PathBuf>) -> Result<()> {
    let mut children = fs::read_dir(directory)
        .with_context(|| format!("read {}", directory.display()))?
        .collect::<std::io::Result<Vec<_>>>()?;
    children.sort_by(|left, right| {
        left.file_name()
            .as_encoded_bytes()
            .cmp(right.file_name().as_encoded_bytes())
    });
    for child in children {
        let path = child.path();
        entries.push(path.clone());
        if fs::symlink_metadata(&path)?.is_dir() {
            collect(root, &path, entries)?;
        }
    }
    entries.sort_by(|left, right| {
        let left = left
            .strip_prefix(root)
            .unwrap_or(left)
            .as_os_str()
            .as_encoded_bytes();
        let right = right
            .strip_prefix(root)
            .unwrap_or(right)
            .as_os_str()
            .as_encoded_bytes();
        left.cmp(right)
    });
    Ok(())
}

pub fn export_workspace_file(root: &Path, destination: &Path) -> Result<(u64, String)> {
    let result = (|| {
        let mut paths = Vec::new();
        collect(root, root, &mut paths)?;
        let mut out = SnapshotWriter::create(destination)?;
        out.write_all(MAGIC)?;
        let mut hard_links: HashMap<(u64, u64), String> = HashMap::new();
        let mut buffer = [0u8; 64 * 1024];
        for absolute in paths {
            let relative = validate_relative(absolute.strip_prefix(root)?)?;
            let metadata = fs::symlink_metadata(&absolute)?;
            let mode = metadata.permissions().mode() & 0o7777;
            if mode & 0o6000 != 0 {
                bail!("setuid/setgid entry rejected: {relative}");
            }
            let raw_mtime_ns =
                metadata.mtime() as u64 * 1_000_000_000 + metadata.mtime_nsec() as u64;
            let mtime_ns = ((raw_mtime_ns + 500_000) / 1_000_000) * 1_000_000;
            if metadata.is_dir() {
                write_header(&mut out, &relative, TYPE_DIRECTORY, mode, mtime_ns, 0)?;
            } else if metadata.file_type().is_symlink() {
                let target = fs::read_link(&absolute)?;
                if target.is_absolute()
                    || target
                        .components()
                        .any(|part| matches!(part, Component::ParentDir))
                {
                    bail!("symlink target escapes workspace: {relative}");
                }
                let target = target.to_string_lossy();
                write_header(
                    &mut out,
                    &relative,
                    TYPE_SYMLINK,
                    mode,
                    mtime_ns,
                    target.len() as u64,
                )?;
                out.write_all(target.as_bytes())?;
            } else if metadata.is_file() {
                if metadata.len() > MAX_FILE_BYTES {
                    bail!("file exceeds snapshot limit: {relative}");
                }
                let key = (metadata.dev(), metadata.ino());
                if metadata.nlink() > 1 {
                    if let Some(target) = hard_links.get(&key) {
                        write_header(
                            &mut out,
                            &relative,
                            TYPE_HARDLINK,
                            mode,
                            mtime_ns,
                            target.len() as u64,
                        )?;
                        out.write_all(target.as_bytes())?;
                        continue;
                    }
                    hard_links.insert(key, relative.clone());
                }
                write_header(
                    &mut out,
                    &relative,
                    TYPE_FILE,
                    mode,
                    mtime_ns,
                    metadata.len(),
                )?;
                let mut file = File::open(&absolute)?;
                loop {
                    let count = file.read(&mut buffer)?;
                    if count == 0 {
                        break;
                    }
                    out.write_all(&buffer[..count])?;
                }
            } else {
                bail!("unsupported snapshot entry: {relative}");
            }
        }
        out.finish()
    })();
    if result.is_err() {
        let _ = fs::remove_file(destination);
    }
    result
}
